mountpoint-s3/src/data_cache/multilevel_cache.rs

```diff
@@ -0,0 +1,314 @@
+use std::sync::Arc;
+
+use async_trait::async_trait;
+use futures::task::{Spawn, SpawnExt};
+use tracing::{trace, warn};
+
+use crate::object::ObjectId;
+
+use super::{BlockIndex, ChecksummedBytes, DataCache, DataCacheResult};
+
+pub struct MultilevelDataCache<DiskCache, ExpressCache, Runtime> {
+    disk_cache: Arc<DiskCache>,
+    express_cache: ExpressCache,
+    runtime: Runtime,
+}
+
+impl<DiskCache: DataCache, ExpressCache: DataCache, Runtime: Spawn>
+    MultilevelDataCache<DiskCache, ExpressCache, Runtime>
+{
+    pub fn new(disk_cache: Arc<DiskCache>, express_cache: ExpressCache, runtime: Runtime) -> Self {
+        assert_eq!(disk_cache.block_size(), express_cache.block_size());
+        Self {
+            disk_cache,
+            express_cache,
+            runtime,
+        }
+    }
+}
+
+#[async_trait]
+impl<DiskCache, ExpressCache, Runtime> DataCache for MultilevelDataCache<DiskCache, ExpressCache, Runtime>
+where
+    DiskCache: DataCache + Sync + Send + 'static,
+    ExpressCache: DataCache + Sync,
+    Runtime: Spawn + Sync,
+{
+    async fn get_block(
+        &self,
+        cache_key: &ObjectId,
+        block_idx: BlockIndex,
+        block_offset: u64,
+    ) -> DataCacheResult<Option<ChecksummedBytes>> {
+        match self.disk_cache.get_block(cache_key, block_idx, block_offset).await {
+            Ok(Some(data)) => {
+                trace!(cache_key=?cache_key, block_idx=block_idx, "block served from the disk cache");
+                return DataCacheResult::Ok(Some(data));
+            }
+            Ok(None) => (),
+            Err(err) => warn!(cache_key=?cache_key, block_idx=block_idx, ?err, "error reading block from disk cache"),
+        }
+
+        if let Some(data) = self.express_cache.get_block(cache_key, block_idx, block_offset).await? {
+            trace!(cache_key=?cache_key, block_idx=block_idx, "block served from the express cache");
+            let cache_key = cache_key.clone();
+            let cache = self.disk_cache.clone();
+            let data_cloned = data.clone();
```

Comment:
In the future, do you think this would be worth wrapping in an Arc to avoid a clone?

Reply:
`ChecksummedBytes` is cheaply cloneable  (it's size is about 7 `usize` fields), so I think it's ok to pass the value around.